README.md
Guanyin or Guan Yin is an East Asian bodhisattva associated with compassion as venerated by Mahayana Buddhists. She is commonly known as the "Goddess of Mercy" in English. The Chinese name Guanyin is short for Guanshiyin, meaning "[The 
One Who] Perceives the Sounds of the World". (from https://en.wikipedia.org/wiki/Guanyin)

# Getting started

## Requirements

    Node v14.x or higher (recommended Node v18)
    NPM (comes with Node)
    PostgreSQL 10 or higher (developed on 10.1)
    Docker (if using for database migrations)

Checking for node:

    node -v


## Installing modules

    $ npm install

## Postgres setup

### Setting environment variables

Create a `.env` file and populate it. The `.env-example` file provides a template for this.

### Create a PostgreSQL database

    $ psql postgres -U postgres

    # create database ${DATABASE};
    # create user ${USER};
    # alter role ${USER} with password '${PASSWORD}';
    # grant all on database ${DATABASE} to ${USER};
    # \q


### Migrating the database

Database migrations can be applied manually, but we like Flyway for applying migrations in a controlled way. To
use Flyway for migrations, when setting up the database for the first time:

Create a file in conf/ called `flyway.conf`. The `conf/example-flyway.conf` file provides a template for this.

Pull in the Flyway Docker image:

    $ docker pull flyway/flyway
    
Perform the initial migration using the following:

    $ npm run fw:migrate

After that initial setup, run migrations as necessary using the same command.

If the database needs to be wiped clean and reset, this can be done using:

    $ npm run fw:clean
    $ npm run fw:migrate

Note that the argument `--network=host` in `package.json`'s `fw:clean` and `fw:migrate` are particularly important if 
`flyway.url` includes `localhost`.

## Running the application

    $ npm run start

### Testing
Pull docker images for postgres and flyway

    $ docker pull postgres:15
    $ docker pull flyway/flyway

Run tests using:

    $ npm run test
    
To wipe clean and reset the docker container for the test database without running the tests use:

    $ npm run pretest

## Building and Running using Docker
A container can be built using Docker. You will need an additional container
running PostgreSQL.

    docker build --target flyway-migration -t "oicrgsi/guanyin-flyway:${version}" .
    docker build --target webapp -t "oicrgsi/guanyin:${version}" .

To run, different environment variables are needed for each container. For the
migration container, set the [Flyway environment
variables](https://flywaydb.org/documentation/envvars) to connect to your
database. For the webapp container, set `DB_CONNECTION` to point to the
PostgreSQL database.

     docker run -e FLYWAY_USER=reportdb -e FLYWAY_PASSWORD=secret FLYWAY_URL=jdbc:postgresql://localhost/reportdb -t oicrgsi/guanyin-flyway:${version}
     docker run -p 3000:3000 -e DB_CONNECTION=postgresql://reportdb:secret @localhost/reportdb -t oicrgsi/guanyin:${version}


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and as of version 1.0.0, follows semantic versioning.

## Unreleased
### Added
  * Support for Node v14+ (Recommended Node v18)

### Removed
  * Support for Node v12 and earlier


## [1.4.0] 2021-10-28
### Added
  * Search report records by report ID

### Changed
  * Use streaming for getAllReports and getAllReportrecords

## [1.3.0] 2021-04-09
### Added
  * Increased allowable request body size
  * Added logging of which Shesmu parameters fail validation

## [1.2.0] 2021-04-09
### Added
  * Allow deleting reports without report records
  * Allow building using Docker

## [1.1.0]  2019-02-21
### Added:
  * Prometheus monitoring
  * Logging
  * API data validation
  * API endpoint for creating minimal report records and filling in details later
  * Support for Shesmu paths, objects
### Changed:
  * Updated API endpoints
    * Removed `/reportdb/record_notification`
      * Replaced with `/reportdb/record/{report_record_id}/notification`
    * Added ability to search for report records with matching report ID and parameters
  * Tests now run using Flyway Docker container


test/setup_test_db.sh
#!/bin/bash

# source the .env file from the test folder
# source might be necessary depending on system
set -a; . "$(pwd)/test/.env";  source "$(pwd)/test/.env"; set +a

# since we need to mount all the migrations in a single directory, the "regular" migrations need to be
# copied into the test folder, then deleted after.
cp sql/V*.sql test/sql/


# create flyway.conf file
echo "flyway.url=jdbc:postgresql://${DB_HOST}:${DB_PORT}/${DB_NAME}" > test/flyway.conf
echo "flyway.user=${DB_USER}" >> test/flyway.conf
echo "flyway.password=${DB_PW}" >> test/flyway.conf
echo "flyway.cleanDisabled=false" >> test/flyway.conf

# stop and remove any previously-running test containers
docker stop "${PG_DB_CONTAINER_NAME}" || true  # it's fine if the container isn't running
docker rm "${PG_DB_CONTAINER_NAME}"

# set up the Postgres database
docker run --name "${PG_DB_CONTAINER_NAME}" -e POSTGRES_USER=$DB_USER -e POSTGRES_PASSWORD=$DB_PW -e POSTGRES_DB=$DB_NAME -d -p $DB_PORT:5432 postgres:15

# give docker time to setup the container before connecting
sleep 1

# perform cleaning and database migration
docker run --rm -v $(pwd)/test:/flyway/conf -v $(pwd)/sql:/flyway/sql --network=host flyway/flyway clean && \
docker run --rm -v $(pwd)/test:/flyway/conf -v $(pwd)/test/sql:/flyway/sql --network=host flyway/flyway migrate

find test/sql/ -type f ! -name 'V999*.sql' -delete


test/stop_test_dbs.sh
#!/bin/bash

. test/.env

docker stop "${PG_DB_CONTAINER_NAME}"
rm $(pwd)/test/flyway.conf


app.js
'use strict';

require('dotenv').config();
var express = require('express');
var path = require('path');
var morgan = require('morgan');
var favicon = require('serve-favicon');
const logger = require('./utils/logger');
var cookieParser = require('cookie-parser');
var bodyParser = require('body-parser');
const uid = require('uid').uid; // generates a unique ID for each request
const ignoreFrom = process.env.IGNORE_ADDRESS || ''; // to skip logging of requests from IT's security tests
const prometheus = require('prom-client');
prometheus.collectDefaultMetrics({ timeout: 5000 });
const httpRequestDurationMilliseconds = new prometheus.Histogram({
  name: 'http_request_duration_ms',
  help: 'Duration of HTTP requests in ms',
  labelNames: ['route'],
  buckets: [0.1, 5, 15, 50, 100, 200, 300, 400, 500]
});

const httpRequestCounter = new prometheus.Counter({
  name: 'http_request_counter',
  help: 'Number of requests for this endpoint',
  labelNames: ['route', 'method', 'status']
});

var index = require('./routes/index');

var swaggerJSDoc = require('swagger-jsdoc');

var app = express();

// swagger definition
var swaggerDefinition = {
  info: {
    title: 'ReportDB API',
    version: '1.0.0',
    description: 'API for the ReportDB project'
  },
  basePath: '/'
};

// options for the swagger docs
var options = {
  // import swaggerDefinitions
  swaggerDefinition: swaggerDefinition,
  // path to the API docs
  apis: ['./routes/*.js']
};

// initialize swagger-jsdoc
var swaggerSpec = swaggerJSDoc(options);

// serve swagger
app.get('/swagger.json', function(req, res) {
  res.setHeader('Content-Type', 'application/json');
  res.send(swaggerSpec);
});

// view engine setup
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'jade');

app.use(morgan('common'));

app.use(bodyParser.json({ limit: '500mb' }));
app.use(bodyParser.urlencoded({ extended: false }));
app.use(cookieParser());
app.use(express.static(path.join(__dirname, 'public')));
app.use(favicon(path.join(__dirname, 'public', 'favicon.ico')));

app.use((req, res, next) => {
  // have to manually set this because there's no guarantee it'll be called this in future versions of Express
  req._startTime = new Date();
  // generate a unique identifier for each request, if one hasn't already been set
  if (!req.uid) req.uid = uid();
  res.uid = req.uid;
  if (
    req.connection.remoteAddress != ignoreFrom &&
    req.originalUrl != '/metrics'
  ) {
    logger.info({
      uid: req.uid,
      method: req.method,
      url: req.originalUrl,
      origin: req.connection.remoteAddress
    });
  }
  next();
});

app.use('/', index);
app.get('/metrics', async (req, res) => {
  res.set('Content-Type', prometheus.register.contentType);
  res.end(prometheus.register.metrics());
});

// catch 404 for all other endpoints and forward to error handler
app.use(function(req, res, next) {
  var err = new Error('Not Found');
  err.status = 404;
  next(err);
});
app.use((req, res, next) => {
  // log metrics after every request
  if (req.connection.remoteAddress != ignoreFrom) {
    const responseTimeInMs = Date.now() - Date.parse(req._startTime);
    const path = req.route ? req.route.path : req.originalUrl;
    httpRequestDurationMilliseconds.labels(path).observe(responseTimeInMs);
    httpRequestCounter.labels(path, req.method, res.statusCode).inc();
  }
  next();
});

app.use(function(err, req, res, next) {
  if (res._headerSent) {
    // we've already sent a 404 or other error code, which will already be logged
    next();
    return;
  }
  if (!err.status) {
    if (err.isBoom) {
      res
        .status(err.output.statusCode)
        .json({ message: err.data[0].message, path: err.data[0].path });
    } else if (err.name == 'ValidationError') {
      res.status(400).json({
        status: 'error',
        message: err.message
      });
    } else {
      // unexpected error, so log it
      logger.debug(err);
      logger.error({ url: req.originalUrl, message: err.message });
      res.status(500).json({
        status: 'error',
        message: app.get('env') === 'production' ? 'system error' : err //return err only in development env.
      });
    }
  } else {
    logger.debug(err);
    logger.error({ url: req.originalUrl, message: err.message });
    res.status(err.status).json({
      status: 'error',
      message: err.message
    });
  }
  res.end();
  next();
});

module.exports = app;


utils/logger.js


'use strict';

const winston = require('winston');
const logLocation = process.env.LOG_LOCATION || 'logs';

const logger = winston.createLogger();
if (process.env.NODE_ENV !== 'test'){ 
  logger.add({
    transports: [
      new winston.transports.File({
        name: 'combined-log',
        filename: `${logLocation}/combined.log`,
        level: 'info',
        colorize: true,
        timestamp: 'tsFormat'
      }),
      new winston.transports.File({
        name: 'error-log',
        filename: `${logLocation}/error.log`,
        level: 'error',
        handleException: true,
        humanReadableUnhandledException: true,
        colorize: true,
        timestamp: 'tsFormat'
      })
    ]
  });
}

if (process.env.NODE_ENV !== 'production' || process.env.LOG_LEVEL != 'prod') {
  logger.add(new winston.transports.Console({
    level: 'debug',
    colorize: true,
    timestamp: 'tsFormat'
  }));
}

module.exports = logger;

utils/express-joi-validator.js
'use strict';

const Joi = require('joi');

module.exports = function validate(schema, options) {
  options = options || {
    abortEarly: false
  };

  function ValidationError(message, details) {
    this.status = 400;
    this.message = message;
    if (details) this.details = details;
  }
  ValidationError.prototype = Error.prototype;

  return function validateRequest(req, res, next) {
    const toValidate = {};
    if (!schema) return next();

    ['params', 'body', 'query'].forEach(key => {
      if (schema.describe().keys[key] !== undefined) {
        toValidate[key] = req[key];
      }
    });

   const { value, error } = schema.validate(toValidate, options);
 
   if (error) return next(new ValidationError(error.message, error.details));
   // copy the validated data to the req object
   Object.assign(req, value);

   return next();
  };
};


queries.js
const promise = require('bluebird');
const QueryStream = require('pg-query-stream');
const JSONStream = require('JSONStream');

const options = {
  // Initialization Options
  promiseLib: promise
};

const pgp = require('pg-promise')(options);
const db = pgp(process.env.DB_CONNECTION);

const logger = require('./utils/logger');

/**
 * Parse a Shesmu type signature and return an object that can check and sort values.
 *
 * This returns an object with three properties: check (checks if object
 * matches signature), compare (compares two values and determines canonicalise
 * order), canonicalise ("sort" a value), rest (the unparsed type string; this
 * should be empty for a valid string). Null is returned if the type is not
 * valid.
 */
function parseShesmuType(type) {
  switch (type.charAt(0)) {
    case 'i':
    case 'd': // This is intentional; dates are millis since the epoch
      return {
        check: value => typeof value === 'number',
        compare: (a, b) => a - b,
        canonicalise: value => value,
        rest: type.substring(1)
      };
    case 's':
    case 'p': // This is intentional; paths are still strings
      return {
        check: value => typeof value === 'string',
        compare: (a, b) => a.localeCompare(b),
        canonicalise: value => value,
        rest: type.substring(1)
      };
    case 'b':
      return {
        check: value => typeof value === 'boolean',
        compare: (a, b) => a - b,
        canonicalise: value => value,
        rest: type.substring(1)
      };
    case 'a':
      return (() => {
        const inner = parseShesmuType(type.substr(1));
        if (inner == null) {
          return null;
        }
        return {
          check: value =>
            Array.isArray(value) && value.every(child => inner.check(child)),
          compare: (a, b) =>
            a.length - b.length ||
            a.reduce(
              (acc, avalue, index) => acc || inner.compare(avalue, b[index]),
              0
            ),
          canonicalise: value =>
            value.map(inner.canonicalise).sort(inner.compare),
          rest: inner.rest
        };
      })();
    case 't':
    case 'o':
      return (() => {
        let match;
        if ((match = /^([0-9]*)([^0-9].*)$/.exec(type.substr(1))) === null) {
          return null;
        }
        let rest = match[2];
        const count = parseInt(match[1]);
        if (type.charAt(0) == 't') {
          const types = [];
          for (let index = 0; index < count; index++) {
            const type = parseShesmuType(rest);
            if (type == null) {
              return null;
            }
            rest = type.rest;
            types.push(type);
          }
          return {
            check: value =>
              Array.isArray(value) &&
              value.length == types.length &&
              value.every((child, index) => types[index].check(child)),
            compare: (a, b) =>
              types.reduce(
                (acc, type, index) => acc || type.compare(a[index], b[index]),
                0
              ),
            canonicalise: value =>
              value.map((child, index) => types[index].canonicalise(child)),
            rest: rest
          };
        } else {
          const fields = [];
          for (let index = 0; index < count; index++) {
            if ((match = /^([^$]*)\$(.*)$/.exec(rest)) === null) {
              return null;
            }
            rest = match[2];
            const type = parseShesmuType(rest);
            if (type == null) {
              return null;
            }
            rest = type.rest;
            fields.push({ name: match[1], type: type });
          }
          return {
            check: value =>
              value !== null &&
              typeof value == 'object' &&
              !Array.isArray(value) &&
              Object.keys(value).length == fields.length &&
              fields.every(field => {
                const validField =
                  value.hasOwnProperty(field.name) &&
                  field.type.check(value[field.name]);
                if (!validField)
                  console.warn(
                    'Failed for ' + field.name + ': ' + JSON.stringify(value)
                  );
                return validField;
              }),
            compare: (a, b) =>
              fields.reduce(
                (acc, field) =>
                  acc || field.type.compare(a[field.name], b[field.name]),
                0
              ),
            canonicalise: value => {
              const result = {};
              fields.forEach(
                field =>
                  (result[field.name] = field.type.canonicalise(
                    value[field.name]
                  ))
              );

              return result;
            },
            rest: rest
          };
        }
      })();
    default:
      return null;
  }
}

/** set up custom error if bad params are given */
function ValidationError(message) {
  this.name = 'ValidationError';
  this.message = message || '';
}
ValidationError.prototype = Error.prototype;

/**
 * validate parameters against permitted parameters
 */
function throwIfHasInvalidParams(req, permitted_parameters) {
  const permittedNames = Object.keys(permitted_parameters);
  for (let nameIndex = 0; nameIndex < permittedNames.length; nameIndex++) {
    const name = permittedNames[nameIndex];
    if (req.body.parameters.hasOwnProperty(name)) {
      let type = parseShesmuType(permitted_parameters[name].type);
      if (!type.check(req.body.parameters[name])) {
        throw new ValidationError(
          'Invalid value for parameter ' +
            name +
            '. Expected type ' +
            permitted_parameters[name].type +
            '.'
        );
      }
      req.body.parameters[name] = type.canonicalise(req.body.parameters[name]);
    } else if (permitted_parameters[name].required) {
      throw new ValidationError(
        'Report requires parameter "' + name + '", but it was not supplied.'
      );
    }
  }
  const providedNames = Object.keys(req.body.parameters);
  for (let nameIndex = 0; nameIndex < providedNames.length; nameIndex++) {
    const name = providedNames[nameIndex];
    if (!permitted_parameters.hasOwnProperty(name)) {
      throw new ValidationError('Invalid parameter "' + name + '" provided.');
    }
  }
  return 1;
}

// add query functions

async function getAllReports(req, res, next) {
  const query = new QueryStream('select * from report');
  try {
    const streamed = await db.stream(query, stream => {
      res.status(200);
      res.setHeader('Content-Type', 'application/json');
      stream.pipe(JSONStream.stringify()).pipe(res);
    });
    logger.info({
      streamRowsProcessed: streamed.processed,
      streamingDuration: streamed.duration,
      method: 'getAllReports'
    });
    next();
  } catch (err) {
    return next(err);
  }
}

async function getSingleReport(req, res, next) {
  try {
    const report = await getReportById(req.params.id);
    return res.status(200).json(report);
  } catch (err) {
    returnIfNotFound(err, res, 'report');
    return next(err);
  }
}

async function getAllreports_by_name(req, res, next) {
  try {
    const reports = await db.any(
      'select * from report where name = $1',
      req.query.name
    );
    return res.status(200).json(reports);
  } catch (err) {
    return next(err);
  }
}

function checkTypeDefinitionValidity(parameters) {
  return Object.values(parameters).every(p => {
    const type = parseShesmuType(p.type);
    return type != null && type.rest === '';
  });
}

async function createReport(req, res, next) {
  try {
    if (!req.body.hasOwnProperty('lims_entity')) {
      // property may be null if it is not associated with a LIMS entity (Project, Library, Run, Pool)
      req.body.lims_entity = null;
    }
    if (!checkTypeDefinitionValidity(req.body.permitted_parameters)) {
      throw new ValidationError('Invalid type signature on parameter.');
    }
    const reportData = { ...req.body };
    const insert = await db.one(
      'insert into report(name, version, category, permitted_parameters, lims_entity)' +
        'values(${name}, ${version}, ${category}, ${permitted_parameters}, ${lims_entity})' +
        'returning report_id',
      reportData
    );
    return res.status(201).json(insert);
  } catch (err) {
    return next(err);
  }
}

async function getAllReportrecords(req, res, next) {
  const query = new QueryStream('select * from report_record');
  try {
    const streamed = await db.stream(query, stream => {
      res.status(200);
      res.setHeader('Content-Type', 'application/json');
      stream.pipe(JSONStream.stringify()).pipe(res);
    });
    logger.info({
      streamRowsProcessed: streamed.processed,
      streamingDuration: streamed.duration,
      method: 'getAllReportrecords'
    });
    next();
  } catch (err) {
    return next(err);
  }
}

async function getAllreportrecords_by_notification_done(req, res, next) {
  try {
    const records = await db.any(
      'select * from report_record where notification_done = $1',
      req.query.notification_done
    );
    return res.status(200).json(records);
  } catch (err) {
    return next(err);
  }
}

async function getAllReportrecordsByReportId(req, res, next) {
  try {
    const records = await db.any(
      'select * from report_record where report_id = $1',
      req.params.id
    );
    return res.status(200).json(records);
  } catch (err) {
    return next(err);
  }
}

async function getSingleReportrecord(req, res, next) {
  try {
    const record = await db.one(
      'select * from report_record where report_record_id = $1',
      parseInt(req.params.id)
    );
    return res.status(200).json(record);
  } catch (err) {
    returnIfNotFound(err, res, 'report record');
    return next(err);
  }
}

async function createReportrecord(req, res, next) {
  try {
    const reportID = parseInt(req.body.report_id);
    const report = await getReportById(reportID);
    confirmReportParametersAreValid(req, report, reportID);
    const recordData = { ...req.body };
    recordData.report_id = reportID;
    const insert = await db.one(
      'insert into report_record(report_id, finished, date_generated, freshest_input_date, files_in, report_path, notification_targets, notification_message, parameters)' +
        'values(${report_id}, true, ${date_generated}, ${freshest_input_date}, ${files_in}, ${report_path}, ${notification_targets}, ${notification_message}, ${parameters})' +
        'returning report_record_id',
      recordData
    );
    return res.status(201).json(insert);
  } catch (err) {
    returnIfNotFound(err, res, 'report'); // only getReportById will throw a not found error
    return next(err);
  }
}

async function getReportById(reportID) {
  return await db.one('select * from report where report_id = $1', reportID);
}

function confirmReportParametersAreValid(req, report, reportID) {
  if (report == null)
    throw new ValidationError('Could not find report for ID ' + reportID);
  if (req.body == null || !req.body.hasOwnProperty('parameters'))
    throw new ValidationError(
      'Could not validate parameters because no "parameters" object was provided'
    );
  throwIfHasInvalidParams(req, report.permitted_parameters);
}

async function deleteReport(req, res, next) {
  try {
    if (!req.params.hasOwnProperty('id')) {
      throw new ValidationError('Must include an id of the report to delete');
    }
    const relatedReportRecords = await db.any(
      'select * from report_record where report_id = $1',
      req.params.id
    );
    if (relatedReportRecords.length) {
      throw new ValidationError(
        'Cannot delete report ' +
          req.params.id +
          ' as it has ' +
          relatedReportRecords.length +
          ' associated report record(s)'
      );
    }
    const report = await db.any(
      'select * from report where report_id = $1',
      req.params.id
    );
    if (!report.length) {
      // report was not found
      return res.sendStatus(404);
    }

    await db.result('delete from report where report_id = $1', req.params.id);
    return res.sendStatus(204);
  } catch (err) {
    return next(err);
  }
}

async function createReportrecordStart(req, res, next) {
  try {
    const reportID = parseInt(req.query.report);
    const report = await getReportById(reportID);
    confirmReportParametersAreValid(req, report, reportID);
    const recordData = { ...req.body };
    recordData.report_id = reportID;
    const insert = await db.one(
      'insert into report_record(report_id, finished, date_generated, parameters)' +
        'values(${report_id}, false, NOW(), ${parameters})' +
        'returning report_record_id',
      recordData
    );
    return res.status(201).json(insert);
  } catch (err) {
    return next(err);
  }
}

async function patchReportrecord(req, res, next) {
  try {
    const recordData = { ...req.body };
    recordData.report_record_id = parseInt(req.params.id);
    await db.one(
      'update report_record set finished=true, freshest_input_date=${freshest_input_date}, files_in=${files_in}, report_path=${report_path}, notification_targets=${notification_targets}, notification_message=${notification_message} where report_record_id=${report_record_id} and finished=false returning report_record_id',
      recordData
    );
    return res.status(200).json({
      status: 'success',
      message: 'Updated report record'
    });
  } catch (err) {
    returnIfNotFound(err, res, 'unfinished report record');
    return next(err);
  }
}

async function updateReportrecord_notification_done(req, res, next) {
  try {
    await db.one(
      'update report_record set notification_done=true where report_record_id=$1 returning report_record_id',
      [parseInt(req.params.id)]
    );
    return res.status(200).json({
      status: 'success',
      message: 'Updated report record'
    });
  } catch (err) {
    returnIfNotFound(err, res, 'report record');
    return next(err);
  }
}

async function findReportrecord_files_in(req, res, next) {
  try {
    const record = await db.any(
      'select rr.* from report_record rr, report r  where r.report_id=rr.report_id and r.name = $1 and r.version = $2 and rr.files_in= $3',
      [req.query.name, req.query.version, req.body.files_in]
    );
    return res.status(200).json(record);
  } catch (err) {
    return next(err);
  }
}

async function findReportrecord_parameters(req, res, next) {
  let report;
  try {
    if (!req.body) {
      throw new ValidationError('Parameters must be provided in request body');
    }
    if (!req.body.hasOwnProperty('parameters')) {
      throw new ValidationError('Parameters must be provided in request body');
    }
    if (req.query.hasOwnProperty('report')) {
      report = await getReportById(parseInt(req.query.report));
    } else {
      report = await db.one(
        'select * from report where name = $1 and version = $2',
        [req.query.name, req.query.version]
      );
    }
  } catch (err) {
    returnIfNotFound(err, res, 'report');
    return next(err);
  }
  if (report == null || report.report_id == null) {
    send404(res, 'report');
  }
  try {
    confirmReportParametersAreValid(req, report, report.report_id);
    const record = await db.any(
      'select rr.* from report_record rr where rr.report_id=$1 and rr.parameters= $2',
      [report.report_id, req.body.parameters]
    );
    return res.status(200).json(record);
  } catch (err) {
    returnIfNotFound(err, res, 'report record');
    return next(err);
  }
}

/** Catches errors thrown by `db.one` calls which returned no items */
function returnIfNotFound(err, res, itemName) {
  if (err.result != null && err.result.rowCount == 0) {
    send404(res, itemName);
  }
}

function send404(res, itemName) {
  return res.status(404).json({
    status: 'not found',
    message: `The ${itemName} you requested could not be found`
  });
}

module.exports = {
  getAllReports: getAllReports,
  getSingleReport: getSingleReport,
  getAllreports_by_name: getAllreports_by_name,
  createReport: createReport,
  deleteReport: deleteReport,
  getAllReportrecords: getAllReportrecords,
  getAllReportrecordsByReportId: getAllReportrecordsByReportId,
  getSingleReportrecord: getSingleReportrecord,
  createReportrecord: createReportrecord,
  createReportrecordStart: createReportrecordStart,
  patchReportrecord: patchReportrecord,
  updateReportrecord_notification_done: updateReportrecord_notification_done,
  getAllreportrecords_by_notification_done: getAllreportrecords_by_notification_done,
  findReportrecord_files_in: findReportrecord_files_in,
  findReportrecord_parameters: findReportrecord_parameters
};
